cmd/openshift-tests/e2e.go

```diff
@@ -821,3 +823,65 @@ func getEventsByPod(events []*monitor.EventInterval) map[string][]*monitor.Event
 	}
 	return eventsByPods
 }
+
+func testOperatorStateTransitions(events []*monitor.EventInterval) []*ginkgo.JUnitTestCase {
+	// if this starts tripping too much, or tripping on the same actors, feel free to split into tests
+	// for every operator with independent pass/fails
+	available := testOperatorState("Available", events)
+	degraded := testOperatorState("Degraded", events)
+	progressing := testOperatorState("Progressing", events)
+
+	ret := []*ginkgo.JUnitTestCase{}
+	ret = append(ret, operatorStateTestCases("[sig-cluster-lifecycle] operators should never go unavailable", available)...)
+	ret = append(ret, operatorStateTestCases("[sig-cluster-lifecycle] operators should never go degraded", degraded)...)
+	ret = append(ret, operatorStateTestCases("[sig-cluster-lifecycle] operators should never go progressing", progressing)...)
+	return ret
+}
+
+func operatorStateTestCases(testName string, failures []string) []*ginkgo.JUnitTestCase {
+	success := &ginkgo.JUnitTestCase{Name: testName}
+
+	if len(failures) == 0 {
+		return []*ginkgo.JUnitTestCase{success}
+	}
+
+	failure := &ginkgo.JUnitTestCase{
+		Name:      testName,
+		SystemOut: strings.Join(failures, "\n"),
+		FailureOutput: &ginkgo.FailureOutput{
+			Output: fmt.Sprintf("%d unexpected clusteroperator state transitions \n\n%v", len(failures), strings.Join(failures, "\n")),
+		},
+	}
+
+	// write a passing test to trigger detection of this issue as a flake. Doing this first to try to see how frequent the issue actually is
+	return []*ginkgo.JUnitTestCase{failure, success}
+}
+
+func testOperatorState(interestingCondition string, events []*monitor.EventInterval) []string {
+	failures := []string{}
+
+	clusterOperatorToEvents := getEventsByOperator(events)
```

Comment:
library-go-based operators will event on ClusterOperator changes, but I'm not sure non-library-go operators will event (consistently) on those or not.

Reply:
> library-go-based operators will event on ClusterOperator changes, but I'm not sure non-library-go operators will event (consistently) on those or not.

this wires to e2e monitor events, not real events.  Also, non-library-go operators could learn.... :)

Reply:
i do think we could add an OLM monitor addon - olm isn't as sophisticated at CO unfortuntaely.